authnz/src/main/java/io/stargate/auth/AuthorizationService.java

```diff
@@ -22,17 +22,69 @@
 
 public interface AuthorizationService {
 
+  /**
+   * Using the provided token will perform pre-authorization where possible, executes the query
+   * provided, and then authorizes the response of the query.
+   *
+   * @param action A {@link QueryBuilder} object to be executed and authorized against a token.
+   * @param token The authenticated token to use for authorization.
+   * @param primaryKeyValues A list of primary key values that will be used in the query and should
+   *     be authorized against the token.
+   * @param tableMetadata The {@link Table} that will be queried against.
+   * @return On success will return the result of the query and otherwise will return an exception
+   *     relating to the failure to authorize.
+   * @throws Exception An exception relating to the failure to authorize.
+   */
   ResultSet authorizedDataRead(
       Callable<ResultSet> action, String token, List<String> primaryKeyValues, Table tableMetadata)
       throws Exception;
 
+  /**
+   * Using the provided token will perform pre-authorization where possible and if successful
+   * executes the query provided.
+   *
+   * @param action A {@link QueryBuilder} object to be executed and authorized against a token.
+   * @param token The authenticated token to use for authorization.
+   * @param primaryKeyValues A list of primary key values that will be used in the query and should
+   *     be authorized against the token.
+   * @param tableMetadata The {@link Table} that will be queried against.
+   * @return On success will return the result of the query and otherwise will return an exception
+   *     relating to the failure to authorize.
+   * @throws Exception An exception relating to the failure to authorize.
+   */
   ResultSet authorizedDataWrite(
       Callable<ResultSet> action, String token, List<String> primaryKeyValues, Table tableMetadata)
       throws Exception;
 
+  /**
+   * Using the provided token will perform pre-authorization where possible, executes the query
+   * provided, and then authorizes the response of the query.
+   *
+   * @param action A {@link QueryBuilder} object to be executed and authorized against a token.
+   * @param token The authenticated token to use for authorization.
+   * @param keyspace Either the keyspace containing the resource to be read or the actual resource
+   *     being read.
+   * @param table The table within the provided keyspace that is being read.
+   * @return On success will return the result of the query and otherwise will return an exception
+   *     relating to the failure to authorize.
+   * @throws Exception An exception relating to the failure to authorize.
+   */
   ResultSet authorizedSchemaRead(
```

Comment:
What's the difference between "data" read and "schema" read?

Reply:
A "data" read is going to be a `SELECT ...` while a "schema" read will be like a `DESCRIBE ...`. For now "data" reads can be authorized using claims in the JWT while "schema" reads will be authorized by the role and permissions set up in C*. But this does open up to the possibility for more fine grained IAM using something like [OPA](https://www.openpolicyagent.org/) or Redhat Keycloak.